File: frontend/src/js/friends.js
import { requireAuth, logout, getSession, searchUsers, sendFriendRequest, getFriendRequests, acceptFriendRequest, rejectFriendRequest, getFriends, removeFriend, getFriendActivity, getBusinessById } from "./api-client.js";

if (!requireAuth()) {
    throw new Error("Authentication required");
}

const session = getSession();
const userId = session.userId;

const userSearchInput = document.getElementById("user-search");
const userSearchResults = document.getElementById("user-search-results");
const pendingRequestsEl = document.getElementById("pending-requests");
const friendsListEl = document.getElementById("friends-list");
const activityFeedEl = document.getElementById("activity-feed");
const logoutBtn = document.getElementById("logout-btn");

let searchTimeout = null;

async function loadFriendsData() {
    await Promise.all([
        loadPendingRequests(),
        loadFriendsList(),
        loadActivity(),
    ]);
}

async function loadPendingRequests() {
    try {
        const result = await getFriendRequests(userId);
        if (result.status === "success") {
            if (result.pending.length === 0) {
                pendingRequestsEl.innerHTML = '<p class="empty-text">No pending requests</p>';
                return;
            }

            pendingRequestsEl.innerHTML = result.pending.map(r => `
                <div class="request-item" data-request-id="${r.requestId}">
                    <span class="request-username">${r.fromUsername}</span>
                    <div class="request-actions">
                        <button class="accept-btn" data-id="${r.requestId}">Accept</button>
                        <button class="reject-btn" data-id="${r.requestId}">Reject</button>
                    </div>
                </div>
            `).join("");

            pendingRequestsEl.querySelectorAll(".accept-btn").forEach(btn => {
                btn.addEventListener("click", async () => {
                    const result = await acceptFriendRequest(parseInt(btn.dataset.id), userId);
                    if (result.status === "success") {
                        loadFriendsData();
                    }
                });
            });

            pendingRequestsEl.querySelectorAll(".reject-btn").forEach(btn => {
                btn.addEventListener("click", async () => {
                    const result = await rejectFriendRequest(parseInt(btn.dataset.id), userId);
                    if (result.status === "success") {
                        loadFriendsData();
                    }
                });
            });
        }
    } catch (error) {
        console.error("Error loading requests:", error);
    }
}

async function loadFriendsList() {
    try {
        const result = await getFriends(userId);
        if (result.status === "success") {
            if (result.friends.length === 0) {
                friendsListEl.innerHTML = '<p class="empty-text">No friends yet. Search for users above.</p>';
                return;
            }

            friendsListEl.innerHTML = result.friends.map(f => `
                <div class="friend-item" data-friendship-id="${f.friendshipId}">
                    <div class="friend-info">
                        <span class="friend-username">${f.friendUsername}</span>
                        <span class="friend-since">Friends since ${new Date(f.since).toLocaleDateString()}</span>
                    </div>
                    <button class="remove-friend-btn" data-id="${f.friendshipId}">Remove</button>
                </div>
            `).join("");

            friendsListEl.querySelectorAll(".remove-friend-btn").forEach(btn => {
                btn.addEventListener("click", async () => {
                    const result = await removeFriend(parseInt(btn.dataset.id), userId);
                    if (result.status === "success") {
                        loadFriendsData();
                    }
                });
            });
        }
    } catch (error) {
        console.error("Error loading friends:", error);
    }
}

async function loadActivity() {
    try {
        const result = await getFriendActivity(userId);
        if (result.status === "success") {
            if (result.activity.length === 0) {
                activityFeedEl.innerHTML = '<p class="empty-text">No recent activity from friends.</p>';
                return;
            }

            activityFeedEl.innerHTML = "";
            for (const item of result.activity) {
                let businessName = `Business #${item.businessID}`;
                try {
                    const biz = await getBusinessById(item.businessID);
                    if (biz.status === "success" && biz.business) {
                        businessName = biz.business.name;
                    }
                } catch (e) { /* use default */ }

                const stars = "★".repeat(item.rating) + "☆".repeat(5 - item.rating);
                const date = new Date(item.createdAt).toLocaleDateString();

                const activityItem = document.createElement("div");
                activityItem.className = "activity-item";
                activityItem.innerHTML = `
                    <div class="activity-header">
                        <span class="activity-username">${item.username}</span>
                        <span class="activity-date">${date}</span>
                    </div>
                    <div class="activity-business">${businessName}</div>
                    <div class="activity-rating">${stars}</div>
                    <div class="activity-content">${item.review || ""}</div>
                `;
                activityFeedEl.appendChild(activityItem);
            }
        }
    } catch (error) {
        console.error("Error loading activity:", error);
    }
}

document.addEventListener("DOMContentLoaded", () => {
    loadFriendsData();

    if (logoutBtn) {
        logoutBtn.addEventListener("click", (e) => {
            e.preventDefault();
            logout();
        });
    }

    // User search
    userSearchInput.addEventListener("input", () => {
        clearTimeout(searchTimeout);
        const query = userSearchInput.value.trim();

        if (query.length < 2) {
            userSearchResults.classList.remove("active");
            return;
        }

        searchTimeout = setTimeout(async () => {
            try {
                const result = await searchUsers(query, userId);
                if (result.status === "success" && result.users.length > 0) {
                    userSearchResults.innerHTML = result.users.map(u => `
                        <div class="search-result-item">
                            <span>${u.username}</span>
                            <button class="add-friend-btn" data-user-id="${u.id}">Add Friend</button>
                        </div>
                    `).join("");
                    userSearchResults.classList.add("active");

                    userSearchResults.querySelectorAll(".add-friend-btn").forEach(btn => {
                        btn.addEventListener("click", async () => {
                            const result = await sendFriendRequest(userId, parseInt(btn.dataset.userId));
                            if (result.status === "success") {
                                btn.textContent = "Sent";
                                btn.disabled = true;
                                btn.style.background = "#ccc";
                            } else {
                                alert(result.message);
                            }
                        });
                    });
                } else {
                    userSearchResults.innerHTML = '<div class="search-result-item"><span>No users found</span></div>';
                    userSearchResults.classList.add("active");
                }
            } catch (error) {
                console.error("Search error:", error);
            }
        }, 300);
    });

    // Close search results when clicking outside
    document.addEventListener("click", (e) => {
        if (!e.target.closest(".search-section")) {
            userSearchResults.classList.remove("active");
        }
    });
});

File: backend/api/routes/friends.py
"""
./backend/api/routes/friends.py

API routes for friend relationships and social features.
"""

from flask import Blueprint, jsonify, request

from backend.core import friend_manager as fm
from backend.core import user_manager as um

friends_bp = Blueprint("friends", __name__, url_prefix="/api/friends")


@friends_bp.route("/request", methods=["POST"])
def send_request():
    data = request.get_json()
    if not data:
        return jsonify({"status": "error", "message": "No data provided"}), 400

    from_user_id = data.get("fromUserId")
    to_user_id = data.get("toUserId")

    if not from_user_id or not to_user_id:
        return jsonify({"status": "error", "message": "fromUserId and toUserId are required"}), 400

    result = fm.send_friend_request(from_user_id, to_user_id)
    if result["status"] == "error":
        return jsonify(result), 400
    return jsonify(result), 201


@friends_bp.route("/requests", methods=["GET"])
def get_requests():
    user_id = request.args.get("user_id", type=int)
    if not user_id:
        return jsonify({"status": "error", "message": "user_id is required"}), 400

    pending = fm.get_pending_requests(user_id)
    sent = fm.get_sent_requests(user_id)

    # Enrich with usernames
    for r in pending:
        user = um.get_user_by_id(r["fromUserId"])
        r["fromUsername"] = user.get("username", "Unknown") if user else "Unknown"

    for r in sent:
        user = um.get_user_by_id(r["toUserId"])
        r["toUsername"] = user.get("username", "Unknown") if user else "Unknown"

    return jsonify({"status": "success", "pending": pending, "sent": sent})


@friends_bp.route("/requests/<int:request_id>/accept", methods=["PUT"])
def accept_request(request_id):
    data = request.get_json()
    if not data or not data.get("userId"):
        return jsonify({"status": "error", "message": "userId is required"}), 400

    result = fm.accept_request(request_id, data["userId"])
    if result["status"] == "error":
        return jsonify(result), 400
    return jsonify(result)


@friends_bp.route("/requests/<int:request_id>/reject", methods=["PUT"])
def reject_request(request_id):
    data = request.get_json()
    if not data or not data.get("userId"):
        return jsonify({"status": "error", "message": "userId is required"}), 400

    result = fm.reject_request(request_id, data["userId"])
    if result["status"] == "error":
        return jsonify(result), 400
    return jsonify(result)


@friends_bp.route("", methods=["GET"])
def get_friends():
    user_id = request.args.get("user_id", type=int)
    if not user_id:
        return jsonify({"status": "error", "message": "user_id is required"}), 400

    friends = fm.get_friends(user_id)

    # Enrich with usernames
    for f in friends:
        user = um.get_user_by_id(f["friendUserId"])
        f["friendUsername"] = user.get("username", "Unknown") if user else "Unknown"

    return jsonify({"status": "success", "friends": friends})


@friends_bp.route("/<int:friendship_id>", methods=["DELETE"])
def remove_friend(friendship_id):
    user_id = request.args.get("user_id", type=int)
    if not user_id:
        return jsonify({"status": "error", "message": "user_id is required"}), 400

    result = fm.remove_friend(friendship_id, user_id)
    if result["status"] == "error":
        return jsonify(result), 400
    return jsonify(result)


@friends_bp.route("/activity", methods=["GET"])
def get_activity():
    user_id = request.args.get("user_id", type=int)
    if not user_id:
        return jsonify({"status": "error", "message": "user_id is required"}), 400

    activity = fm.get_friend_activity(user_id)
    return jsonify({"status": "success", "activity": activity})


@friends_bp.route("/search", methods=["GET"])
def search_users():
    query = request.args.get("q", "")
    user_id = request.args.get("user_id", type=int)

    if not query or len(query) < 2:
        return jsonify({"status": "error", "message": "Query must be at least 2 characters"}), 400

    users = um.search_users(query)

    # Filter out the requesting user
    if user_id:
        users = [u for u in users if u.get("id") != user_id]

    # Return only safe fields
    safe_users = [{"id": u["id"], "username": u["username"]} for u in users]
    return jsonify({"status": "success", "users": safe_users})
